moveit_ros/moveit_servo/include/moveit_servo/utils/common.hpp

```diff
@@ -0,0 +1,158 @@
+ * @return True if the command is valid, else False.
+ */
+bool isValidCommand(const TwistCommand& command);
+
+/**
+ * \brief Checks if a given Pose command is valid.
+ * @param command The command to be checked.
+ * @return True if the command is valid, else False.
+ */
+bool isValidCommand(const PoseCommand& command);
+
+/**
+ * \brief Create a pose message for the provided change in Cartesian position.
+ * @param delta_x The change in Cartesian position.
+ * @param base_to_tip_frame_transform The transformation from robot base to ee frame.
+ * @return The pose message.
+ */
+geometry_msgs::msg::Pose poseFromCartesianDelta(const Eigen::VectorXd& delta_x,
+                                                const Eigen::Isometry3d& base_to_tip_frame_transform);
+
+/**
+ * \brief Create a trajectory message from given joint state.
+ * @param servo_params The configuration used by servo, required for setting some field of the trajectory message.
+ * @param joint_state The joint state to be added into the trajectory.
+ * @return The trajectory message.
+ */
+trajectory_msgs::msg::JointTrajectory composeTrajectoryMessage(const servo::Params& servo_params,
+                                                               const KinematicState& joint_state);
+
+/**
+ * \brief Computes scaling factor for velocity when the robot is near a singularity.
+ * @param joint_model_group The joint model group of the robot, used for fetching the Jacobian.
+ * @param current_state The current state of the robot, used for singularity look ahead.
+ * @param target_delta_x The vector containing the required change in Cartesian position.
+ * @param servo_params The servo parameters, contains the singularity thresholds.
+ * @return The velocity scaling factor and the reason for scaling.
+ */
+std::pair<double, StatusCode> velocityScalingFactorForSingularity(const moveit::core::RobotStatePtr& robot_state,
+                                                                  const Eigen::VectorXd& target_delta_x,
+                                                                  const servo::Params& servo_params);
+
+/**
+ * \brief Apply velocity scaling based on joint limits.
+ * @param velocities The commanded velocities.
+ * @param joint_bounds The bounding information for the robot joints.
+ * @param scaling_override The user defined velocity scaling override.
+ * @return The velocity scaling factor.
+ */
+double jointLimitVelocityScalingFactor(const Eigen::VectorXd& velocities,
+                                       const moveit::core::JointBoundsVector& joint_bounds, double scaling_override);
+
+/**
+ * \brief Finds the joints that are exceeding allowable joint limits.
+ * @param positions The joints positions.
+ * @param velocities The current commanded velocities.
+ * @param joint_bounds The bounding information for the robot joints.
+ * @param margin Additional buffer on the actual joint limits.
+ * @return The joints that are violating the specified position limits.
```

Comment:
```suggestion
 * \brief Finds the joints that are exceeding allowable limits.
 * @param positions The joint positions.
 * @param velocities The current commanded velocities.
 * @param joint_bounds The bounding information for the robot joints.
 * @param margin Additional buffer on the actual joint limits.
 * @return The joints that are violating the specified limits.
```

it's not just position limits that can be exceeded, right?

Reply:
`jointsToHalt` only checks if the commanded velocity will make the joint go past position limit.